bench.py

```diff
@@ -0,0 +1,140 @@
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-x', '--num-epochs', type=int, default=5, help="number of epochs to train for")
+parser.add_argument('-b', '--batch-size', type=int, default=64, help="batch size to train with")
+parser.add_argument('-l', '--block-size', type=int, default=128, help="block size for the model (length of window of context)")
+parser.add_argument('-n', '--num-workers', type=int, default=0, help="number of workers for dataloading")
+parser.add_argument('-g', '--num-gpus', type=int, default=1, help="number of gpus to train on")
+parser.add_argument('-p', '--pin-memory', type=int, default=1, help="pin memory on dataloaders?")
+parser.add_argument('-r', '--precision', type=int, default=32, help="fp precision to use, e.g. 32/16")
+parser.add_argument('-o', '--default_root_dir', type=str, default='.', help="best model checkpoint will be written at this location")
+args = parser.parse_args()
+print(vars(args))
+
+logging.info("preparing the data loaders")
+# NOTE: REDUCED DATA SIZE FOR DEBUGGING, TODO CLEAN BEFORE MERGE IF EVER
+train_dataset = Text8Dataset('text8', args.block_size, crop=(0,         int(1e6)))
+val_dataset   = Text8Dataset('text8', args.block_size, crop=(int(90e6), int(1e5)), override_vocab=train_dataset.vocab)
+test_dataset  = Text8Dataset('text8', args.block_size, crop=(int(95e6), int(1e5)), override_vocab=train_dataset.vocab)
+common = {'batch_size': args.batch_size, 'pin_memory': bool(args.pin_memory), 'num_workers': args.num_workers}
+train_dataloader  = DataLoader(train_dataset, shuffle=True, **common)
+val_dataloader  = DataLoader(val_dataset, shuffle=False, **common)
+test_dataloader  = DataLoader(test_dataset, shuffle=False, **common)
+
+logging.info("creating the model")
+model = GPT(train_dataset.vocab_size, args.block_size, n_layer=6, n_head=8, n_embd=256)
+
+logging.info("preparing the learning rate schedule")
+iter_tokens = args.batch_size * args.block_size # number of tokens backpropped in one iteration
+epoch_tokens = math.ceil(len(train_dataset) / args.batch_size) * iter_tokens
+lr_decay = WarmupCosineLearningRateDecay(learning_rate=6e-4, warmup_tokens=epoch_tokens//2,
+                                         final_tokens=args.num_epochs*epoch_tokens)
+
+t0 = time.time()
+logging.info("training...")
+trainer = pl.Trainer(gpus=args.num_gpus, max_epochs=args.num_epochs, gradient_clip_val=1.0, callbacks=[lr_decay],
+                     precision=args.precision, default_root_dir=args.default_root_dir)
+trainer.fit(model, train_dataloader, val_dataloader)
+t1 = time.time()
+logging.info("%d epochs took %fs, or %fs/epoch", args.num_epochs, t1 - t0, (t1-t0)/args.num_epochs)
+
+# todo below: I don't yet understand the Lightning checkpoint schema
+# logging.info("testing...")
+# ckpt_path = os.path.join(args.default_root_dir, 'model.pt')
+# model.load_from_checkpoint(ckpt_path) # load the best checkpoint we found
+# trainer.test(test_dataloader=test_dataloader)
```

Comment:
```suggestion
# ckpt_path = os.path.join(args.default_root_dir, 'model.pt')
# model.load_from_checkpoint(ckpt_path) # load the best checkpoint we found
# trainer.test(test_dataloader=test_dataloader)
```

```
# Note: LIGHTNING automatically loads the best checkpoint when you call .test()
trainer.test(test_dataloader=test_dataloader)
```

Reply:
got it. it looks like `test_dataloader` is not a kwarg, it's `test_dataloaders` with an 's'. Similar to val_dataloaders, but not the same as train_dataloader without the s, it looks like. Some of the docs are inconsistent on the use of "s" btw, I think.

Reply:
yes. we enable multiple dataloaders for val and test. coming support for train. 
not in research i’m used to, but turns out some people need two datasets to validate haha. go figure

Reply:
some people always need something, which is why frameworks are so hard. Next thing you know you can't use a list of data loaders and have to introduce a `DataLoaderSetManager` object.